src/states/elder/mod.rs

```diff
@@ -714,24 +714,35 @@ impl Elder {
             msg.routing_message()
         );
 
-        let dst = msg.message_dst();
-        let signed_msg = msg.signed_routing_message();
-        if self.in_location(dst) {
-            self.check_signed_message_trust(signed_msg)?;
+        if self.in_location(msg.message_dst()) {
+            let signed_msg = msg.signed_routing_message();
+            let handle = self.check_signed_message_trust(signed_msg)?;
             self.check_signed_message_integrity(signed_msg)?;
-            self.update_our_knowledge(signed_msg);
 
-            if dst.is_multiple() {
-                // Broadcast to the rest of the section.
-                if let Err(error) = self.send_signed_message(&msg) {
-                    debug!("{} Failed to send {:?}: {:?}", self, signed_msg, error);
+            if handle {
+                self.update_our_knowledge(signed_msg);
+                self.routing_msg_queue.push_back(signed_msg.clone());
+            }
+
+            if !msg.message_dst().is_multiple() {
+                if !handle {
+                    // Message is only for us but we can't handle it yet. Backlog it.
+                    self.routing_msg_backlog
+                        .push(msg.into_signed_routing_message());
                 }
+
+                return Ok(());
             }
-            // if addressed to us, then we just queue it and return
-            self.routing_msg_queue
-                .push_back(msg.into_signed_routing_message());
-        } else if let Err(error) = self.send_signed_message(&msg) {
-            debug!("{} Failed to send {:?}: {:?}", self, signed_msg, error);
+        }
+
+        // Relay closer to the destination or broadcast to the rest of our section.
+        if let Err(error) = self.send_signed_message(&msg) {
```

Comment:
Wouldn't moving that before `if self.in_location(msg.message_dst())` simplify the code?

Reply:
No, because we don't always want to do this. For example if the message is untrusted or fails the integrity check.

Reply:
But why shouldn't we? it seem that essentially when receiving a message we want to first send it to everyone that needs receiving it regardless of integrity check/trust since that is what we do for all hops.

That is why I'm thinking having this arbitrary check before resending is probably not necessary, i.e we should not try to avoid resending to all the needed destination regardless of what the processing we do tells us (Also we should probably resend before doing any processing to avoid unnecessary latency).